navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/instruction/DefaultInstructionEngine.java

```diff
@@ -0,0 +1,223 @@
+      @Override
+      public String build(RouteProgress routeProgress) {
+        return buildArrivalInstruction(routeProgress);
+      }
+    });
+  }
+
+  /**
+   * If the next step is greater than 15 meters long, use continue format instruction.
+   * Otherwise, use then in format for instruction
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} to be announced on departure milestone
+   */
+  private String buildDepartureInstruction(RouteProgress progress) {
+    if (progress.getCurrentLegProgress().getUpComingStep().getDistance() > MINIMUM_UPCOMING_STEP_DISTANCE) {
+      return buildContinueFormatInstruction(progress);
+    } else {
+      return buildThenInFormatInstruction(progress);
+    }
+  }
+
+  /**
+   * Create default string format instruction for new step milestone
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} to be announced on new step milestone
+   */
+  private String buildNewStepInstruction(RouteProgress progress) {
+    return buildDefaultFormatInstruction(progress);
+  }
+
+  /**
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} to be announced on imminent milestone
+   */
+  private String buildImminentInstruction(RouteProgress progress) {
+    return buildContinueFormatInstruction(progress);
+  }
+
+  /**
+   * If the next step is less than 15 meters long, use then string format instruction.
+   * Otherwise, just use the upcoming step instruction
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} to be announced on urgent milestone
+   */
+  private String buildUrgentInstruction(RouteProgress progress) {
+    if (progress.getCurrentLegProgress().getUpComingStep().getDistance() < MINIMUM_UPCOMING_STEP_DISTANCE) {
+      return buildThenFormatInstruction(progress);
+    } else {
+      return progress.getCurrentLegProgress().getUpComingStep().getManeuver().getInstruction();
+    }
+  }
+
+  /**
+   * On arrival, use the upcoming step instruction.
+   * If empty, use the current step instruction as a fallback
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} to be announced on departure milestone
+   */
+  private String buildArrivalInstruction(RouteProgress progress) {
+    if (progress.getCurrentLegProgress().getUpComingStep() != null) {
+      String instruction = progress.getCurrentLegProgress().getUpComingStep().getManeuver().getInstruction();
+      if (!TextUtils.isEmpty(instruction)) {
+        return instruction;
+      }
+    } else {
+      return progress.getCurrentLegProgress().getCurrentStep().getManeuver().getInstruction();
+    }
+    return "";
+  }
+
+  /**
+   * Creates a {@link String} with the current step name and distance remaining
+   * Example: "Continue on Main St. for 3.2 miles"
```

Comment:
We should look into better handling the distance values, we might want to further round `3.2 miles` to `3 miles` since the driver most likely won't be able to distinguish the 0.2-mile difference. No need to fix in this PR but opening a ticket might be a good idea.